### PR TITLE
Allow `f` to return a Tuple in `withgradient(f, args...)`

### PR DESCRIPTION
**Pull request by mcabbott**
Maybe we should allow this:
```julia
julia> withgradient([1,2,4]) do x
          z = 1 ./ x
          sum(z), z
       end
(val = (1.75, [1.0, 0.5, 0.25]), grad = ([-1.0, -0.25, -0.0625],))
```
 i.e. return `(val=f(x), grad=gradient(first∘f, x))` when `f` returns a Tuple: Currently an error.

### CODE REVIEW

**Comment by ToucheSir** [comment]
I like the goal but I'm not sure how I feel about the semantics here. Would be nice to get a second opinion on this.

**Comment by mcabbott** [comment]
Maybe this needs an "RFC" label. I agree it's a little weird & potentially surprising. 

Motivation is things like this: https://stackoverflow.com/questions/75921783/accuracy-and-gradient-update-not-within-the-same-training-loop  where it seems cleaner than pushing to some storage, from inside gradient, and hiding that from AD.

**Comment by darsnack** [comment]
We discussed this PR during the community call.

The basic question was: is there another interpretation for the additional elements in an output tuple that isn't auxiliary data? We could consider it from the perspective of `pullback` where multiple sensitivities could be passed in. The final decision was that this interpretation does not meet the definition of a `gradient`, so in the rare case that we do allow multiple sensitivities, we wouldn't call that function `gradient`.

The decision for this PR is to merge it as-is. The only outstanding question was whether `gradient` should mirror the changes to `withgradient` (similar to `jax.grad` and `jax.value_and_grad` linked above).

File: src/compiler/interface.jl [hunk 1]
```diff
@@ -115,12 +115,32 @@ as a named tuple.
 julia> y, ∇ = withgradient(/, 1, 2)
 (val = 0.5, grad = (0.5, -0.25))
 
-julia> ∇ == gradient(/, 1, 2)  # explicit mode
+julia> ∇ == gradient(/, 1, 2)
 true
+```
+
+If `f` returns a Tuple or NamedTuple, then it calculates
+`gradient(first∘f, args...)` but returns the whole `f(args...)`:
```
Comment (CarloLucibello) [inline src/compiler/interface.jl:123]: maybe explain why we allow this, like "Sometimes it is convenient to return some auxiliary outputs"

**Review by ToucheSir — APPROVED** [review]
